### PR TITLE
Provide ace-window-posframe-mode

### PR DESCRIPTION
**Pull request by notmgsk**
Use posframes for overlays.

There does seem to be a slight but noticable delay when displaying the posframes. Haven't investigated too much into why.

### CODE REVIEW

**Comment by abo-abo** [comment]
Thanks. Looks good. Do you have an Emacs Copyright Assignment? Both `ace-window` and `posframe` are on https://elpa.gnu.org/packages/.

**Comment by notmgsk** [comment]
Working on it. Will let you know once it's done.

**Comment by abo-abo** [comment]
Thanks in advance.

**Comment by obar** [comment]
I'm using this with `(set-face-attribute 'aw-leading-char-face nil :weight 'bold :height 2.0)` and the height scale in particular makes the labels stand out. I think it's probably too opinionated to set as a default, but perhaps it warrants inclusion in an example adjacent to the documentation for `ace-window-posframe-mode` in the eventual readme.

This is a useful patch for a great package! Thanks to you both.

File: ace-window-posframe.el [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+         ;; the requested new frame.
+         (bufname (format "*aw-posframe-buffer-%s*" path)))
+    (with-selected-window wnd
+      (push bufname aw--posframe-frames)
+      (posframe-show bufname
+                     :string str
+                     :poshandler aw-posframe-position-handler
+                     :font (face-font 'aw-leading-char-face)
+                     :foreground-color (face-foreground 'aw-leading-char-face)
+                     :background-color (face-background 'aw-leading-char-face)))))
```
Comment (loafofpiecrust) [inline ace-window-posframe.el:23]: I'm excited to have this functionality in `ace-window` mainly for use with `exwm`. For that, though, I need to add the `(parent-frame . nil)` parameter to posframes. So a `defcustom aw-posframe-parameters` would be lovely.

File: ace-window-posframe.el [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      (push bufname aw--posframe-frames)
+      (posframe-show bufname
+                     :string str
+                     :poshandler 'posframe-poshandler-window-center
+                     :font (face-font 'aw-leading-char-face)
+                     :foreground-color (face-foreground 'aw-leading-char-face)
+                     :background-color (face-background 'aw-leading-char-face)))))
+
+(defun aw--remove-leading-chars-posframe ()
+  (map nil #'posframe-delete aw--posframe-frames))
```
Comment (omar-polo) [inline ace-window-posframe.el]: Sorry for my ignorance, but while trying this I got an error about map not being defined.  I don't know which package should provide this (I tried to install and load a package called "map", but it didn't work).  A simpler

    (mapcar #'posframe-delete aw--posframe-frames)

seems to work. (I only did some quick testing).

Another thing is that using your fork, posframe is not listed as dependency (or at least straight.el doesn't think it is -- I am not yet aware of how dependencies are declared and consumed.)

Anyway, thanks for this feature, is awesome :)

**Comment by notmgsk** [comment]
@abo-abo Finally I have the copyright assignment. I'm going to take a look at this again, and address the comments above.

**Comment by gcv** [comment]
@notmgsk: The posframe display delay probably occurs because posframe creation takes a bit of time. You can mitigate the problem by caching and hiding posframes (and lazily creating them as needed, and/or pre-creating a bunch when Emacs idles).

**Comment by notmgsk** [comment]
> @notmgsk: The posframe display delay probably occurs because posframe creation takes a bit of time. You can mitigate the problem by caching and hiding posframes (and lazily creating them as needed, and/or pre-creating a bunch when Emacs idles).

Thanks. The lazy approach is already implemented, and has been sufficient for my usage.

File: ace-window-posframe.el (hunk 1 above)
Comment (tumashu) [reply ace-window-posframe.el:23]: does (parent-frame . nil) work well?

File: ace-window-posframe.el [hunk 3]
```diff
@@ -4,8 +4,15 @@
...
-         (bufname (format "*aw-posframe-buffer-%s*" (gensym))))
+         (str (format "%s" (apply #'string path)))
+         ;; It's important that buffer names are not unique across
+         ;; multiple invocations: posframe becomes very slow when
+         ;; creating new frames, and so being able to reuse old ones
+         ;; makes a huge difference. What defines "able to reuse" is
+         ;; something like: a frame exists which hasn't been deleted
+         ;; (with posframe-delete) and has the same configuration as
+         ;; the requested new frame.
+         (bufname (format "*aw-posframe-buffer-%s*" path)))
```
Comment (tumashu) [inline ace-window-posframe.el:15]: (bufname (format " *aw-posframe-buffer-%s*" path)))  maybe better?

**Comment by josephmturner** [comment]
Hello! Thank you for ace-window! Is there a reason to not merge this PR?

**Comment by josephmturner** [comment]
@abo-abo @notmgsk Thank you!! This will make ace-window work much better with EXWM.

**Pull request closed without merging**